### PR TITLE
Add source code location to all named entities

### LINKED ISSUE #158: [Bug] Format of `textDocument/documentSymbol` LSP response doesn't allegedly match the LSP specification
[Neovim](https://neovim.io/) is a text editor tool which makes use of the Language Server Protocol (LSP) to provide code completion, diagnostics, and other features. The LSP specification defines a response format for the `textDocument/documentSymbol` request, and the one generated by `rust_hdl` seems not be be 100% compliant to the specification.

For example, let consider the following code:

```vhdl
library ieee;
use ieee.std_logic_1164.all;

entity ENTITY_TOP is
  generic (
    GEN : integer := 0
  );
  port (
    INP : in std_logic
  );
end entity;

architecture arch of ENTITY_TOP is
  signal sig : std_logic := '0';

  component ENTITY_1
    generic (
      GEN : integer := 0
    );
    port (
      INP : in std_logic
    );
  end component;

  component ENTITY_2
    generic (
      GEN : integer := 0
    );
    port (
      INP : in std_logic
    );
  end component;

begin

  ENTITY_1_1 : entity work.ENTITY_1(arch1)
  generic map(
    GEN => GEN
  )
  port map(
    INP => INP
  );

  ENTITY_1_2 : entity work.ENTITY_1(arch2)
  generic map(
    GEN => GEN
  )
  port map(
    INP => INP
  );

  ENTITY_2_1 : entity work.ENTITY_2(arch1)
  generic map(
    GEN => GEN
  )
  port map(
    INP => INP
  );

  ENTITY_2_2 : entity work.ENTITY_2(arch2)
  generic map(
    GEN => GEN
  )
  port map(
    INP => INP
  );

  PROC_p: process(INP)
  -----------------------------
  variable var_v : integer := 0;
  -----------------------------
  begin
      -----------------------------
      var_v := 0;
      -----------------------------
      if (INP = '1') then
        sig <= '1';
      else
        sig <= '0';
      end if;
      -----------------------------
  end process;
  -----------------------------

  process
  -----------------------------
  variable var_v : integer := 0;
  -----------------------------
  begin
      -----------------------------
      var_v := 0;
      -----------------------------
      if (INP = '1') then
        sig <= '1';
      else
        sig <= '0';
      end if;
      -----------------------------
  end process;
  -----------------------------
end architecture;
```

In my test, I compared the `textDocument/documentSymbol` response of the following LSP servers:

- `rust_hdl`
- [VHDL-Tool](https://www.vhdltool.com/)
- [vhdl-linter](https://github.com/vhdl-linter/vhdl-linter)

Here is the JSON response to the `textDocument/documentSymbol` request of `rust_hdl`:

```
{
  [7] = {
    result = { {
        children = { {
            kind = 11,
            name = "generic 'GEN'",
            range = {
              ["end"] = {
                character = 7,
                line = 5
              },
              start = {
                character = 4,
                line = 5
              }
            },
            selectionRange = {
              ["end"] = {
                character = 7,
                line = 5
              },
              start = {
                character = 4,
                line = 5
              }
            }
          }, {
            kind = 11,
            name = "port 'INP' : in",
            range = {
              ["end"] = {
                character = 7,
                line = 8
              },
              start = {
                character = 4,
                line = 8
              }
            },
            selectionRange = {
              ["end"] = {
                character = 7,
                line = 8
              },
              start = {
                character = 4,
                line = 8
              }
            }
          }, {
            children = { {
                kind = 24,
                name = "signal 'sig'",
                range = {
                  ["end"] = {
                    character = 12,
                    line = 13
                  },
                  start = {
                    character = 9,
                    line = 13
                  }
                },
                selectionRange = {
                  ["end"] = {
                    character = 12,
                    line = 13
                  },
                  start = {
                    character = 9,
                    line = 13
                  }
                }
              }, {
                children = { {
                    kind = 11,
                    name = "generic 'GEN'",
                    range = {
                      ["end"] = {
                        character = 9,
                        line = 17
                      },
                      start = {
                        character = 6,
                        line = 17
                      }
                    },
                    selectionRange = {
                      ["end"] = {
                        character = 9,
                        line = 17
                      },
                      start = {
                        character = 6,
                        line = 17
                      }
                    }
                  }, {
                    kind = 11,
                    name = "port 'INP' : in",
                    range = {
                      ["end"] = {
                        character = 9,
                        line = 20
                      },
                      start = {
                        character = 6,
                        line = 20
                      }
                    },
                    selectionRange = {
                      ["end"] = {
                        character = 9,
                        line = 20
                      },
                      start = {
                        character = 6,
                        line = 20
                      }
                    }
                  } },
                kind = 5,
                name = "component 'ENTITY_1'",
                range = {
                  ["end"] = {
                    character = 20,
                    line = 15
                  },
                  start = {
                    character = 12,
                    line = 15
                  }
                },
                selectionRange = {
                  ["end"] = {
                    character = 20,
                    line = 15
                  },
                  start = {
                    character = 12,
                    line = 15
                  }
                }
              }, {
                children = { {
                    kind = 11,
                    name = "generic 'GEN'",
                    range = {
                      ["end"] = {
                        character = 9,
                        line = 26
                      },
                      start = {
                        character = 6,
                        line = 26
                      }
                    },
                    selectionRange = {
                      ["end"] = {
                        character = 9,
                        line = 26
                      },
                      start = {
                        character = 6,
                        line = 26
                      }
                    }
                  }, {
                    kind = 11,
                    name = "port 'INP' : in",
                    range = {
                      ["end"] = {
                        character = 9,
                        line = 29
                      },
                      start = {
                        character = 6,
                        line = 29
                      }
                    },
                    selectionRange = {
                      ["end"] = {
                        character = 9,
                        line = 29
                      },
                      start = {
                        character = 6,
                        line = 29
                      }
                    }
                  } },
                kind = 5,
                name = "component 'ENTITY_2'",
                range = {
                  ["end"] = {
                    character = 20,
                    line = 24
                  },
                  start = {
                    character = 12,
                    line = 24
                  }
                },
                selectionRange = {
                  ["end"] = {
                    character = 20,
                    line = 24
                  },
                  start = {
                    character = 12,
                    line = 24
                  }
                }
              }, {
                kind = 2,
                name = "instance 'ENTITY_1_1'",
                range = {
                  ["end"] = {
                    character = 12,
                    line = 35
                  },
                  start = {
                    character = 2,
                    line = 35
                  }
                },
                selectionRange = {
                  ["end"] = {
                    character = 12,
                    line = 35
                  },
                  start = {
                    character = 2,
                    line = 35
                  }
                }
              }, {
                kind = 2,
                name = "instance 'ENTITY_1_2'",
                range = {
                  ["end"] = {
                    character = 12,
                    line = 43
                  },
                  start = {
                    character = 2,
                    line = 43
                  }
                },
                selectionRange = {
                  ["end"] = {
                    character = 12,
                    line = 43
                  },
                  start = {
                    character = 2,
                    line = 43
                  }
                }
              }, {
                kind = 2,
                name = "instance 'ENTITY_2_1'",
                range = {
                  ["end"] = {
                    character = 12,
                    line = 51
                  },
                  start = {
                    character = 2,
                    line = 51
                  }
                },
                selectionRange = {
                  ["end"] = {
                    character = 12,
                    line = 51
                  },
                  start = {
                    character = 2,
                    line = 51
                  }
                }
              }, {
                kind = 2,
                name = "instance 'ENTITY_2_2'",
                range = {
                  ["end"] = {
                    character = 12,
                    line = 59
                  },
                  start = {
                    character = 2,
                    line = 59
                  }
                },
                selectionRange = {
                  ["end"] = {
                    character = 12,
                    line = 59
                  },
                  start = {
                    character = 2,
                    line = 59
                  }
                }
              }, {
                children = { {
                    kind = 13,
                    name = "variable 'var_v'",
                    range = {
                      ["end"] = {
                        character = 16,
                        line = 69
                      },
                      start = {
                        character = 11,
                        line = 69
                      }
                    },
                    selectionRange = {
                      ["end"] = {
                        character = 16,
                        line = 69
                      },
                      start = {
                        character = 11,
                        line = 69
                      }
                    }
                  } },
                kind = 3,
                name = "process 'PROC_p'",
                range = {
                  ["end"] = {
                    character = 8,
                    line = 67
                  },
                  start = {
                    character = 2,
                    line = 67
                  }
                },
                selectionRange = {
                  ["end"] = {
                    character = 8,
                    line = 67
                  },
                  start = {
                    character = 2,
                    line = 67
                  }
                }
              }, {
                kind = 13,
                name = "variable 'var_v'",
                range = {
                  ["end"] = {
                    character = 16,
                    line = 86
                  },
                  start = {
                    character = 11,
                    line = 86
                  }
                },
                selectionRange = {
                  ["end"] = {
                    character = 16,
                    line = 86
                  },
                  start = {
                    character = 11,
                    line = 86
                  }
                }
              } },
            kind = 2,
            name = "architecture 'arch'",
            range = {
              ["end"] = {
                character = 17,
                line = 12
              },
              start = {
                character = 13,
                line = 12
              }
            },
            selectionRange = {
              ["end"] = {
                character = 17,
                line = 12
              },
              start = {
                character = 13,
                line = 12
              }
            }
          } },
        kind = 2,
        name = "entity 'ENTITY_TOP'",
        range = {
          ["end"] = {
            character = 17,
            line = 3
          },
          start = {
            character = 7,
            line = 3
          }
        },
        selectionRange = {
          ["end"] = {
            character = 17,
            line = 3
          },
          start = {
            character = 7,
            line = 3
          }
        }
      } }
  }
}
```

This is the response of the `VHDL-Tool` LSP:

```
{
  [5] = {
    result = { {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 0,
              line = 3
            },
            start = {
              character = 0,
              line = 0
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "imports"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 0,
              line = 1
            },
            start = {
              character = 0,
              line = 0
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "library ieee"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 0,
              line = 3
            },
            start = {
              character = 0,
              line = 1
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "use ieee.std_logic_1164.all"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 0,
              line = 12
            },
            start = {
              character = 0,
              line = 3
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "ENTITY_TOP"
      }, {
        kind = 13,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 9
            },
            start = {
              character = 4,
              line = 8
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "INP"
      }, {
        kind = 14,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 6
            },
            start = {
              character = 4,
              line = 5
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "GEN"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 0,
              line = 101
            },
            start = {
              character = 0,
              line = 12
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "arch"
      }, {
        kind = 13,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 15
            },
            start = {
              character = 2,
              line = 13
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "sig"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 24
            },
            start = {
              character = 2,
              line = 15
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "ENTITY_1"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 0,
              line = 33
            },
            start = {
              character = 2,
              line = 24
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "ENTITY_2"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 43
            },
            start = {
              character = 2,
              line = 35
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "ENTITY_1_1"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 51
            },
            start = {
              character = 2,
              line = 43
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "ENTITY_1_2"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 59
            },
            start = {
              character = 2,
              line = 51
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "ENTITY_2_1"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 67
            },
            start = {
              character = 2,
              line = 59
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "ENTITY_2_2"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 84
            },
            start = {
              character = 2,
              line = 67
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "PROC_p"
      }, {
        kind = 13,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 71
            },
            start = {
              character = 2,
              line = 69
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "var_v"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 0,
              line = 100
            },
            start = {
              character = 2,
              line = 84
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "Unlabelled process"
      }, {
        kind = 13,
        location = {
          range = {
            ["end"] = {
              character = 2,
              line = 88
            },
            start = {
              character = 2,
              line = 86
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "var_v"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 0,
              line = 33
            },
            start = {
              character = 2,
              line = 13
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "head"
      }, {
        kind = 2,
        location = {
          range = {
            ["end"] = {
              character = 0,
              line = 100
            },
            start = {
              character = 2,
              line = 35
            }
          },
          uri = "file:///c:/src_test/ENTITY_TOP.vhd"
        },
        name = "body"
      } }
  }
}
```

And finally the response of the `vhdl-linter` LSP:

```
{
  [5] = {
    result = { {
        detail = "entity",
        kind = 11,
        name = "ENTITY_TOP",
        range = {
          ["end"] = {
            character = 10,
            line = 10
          },
          start = {
            character = 0,
            line = 3
          }
        },
        selectionRange = {
          ["end"] = {
            character = 17,
            line = 3
          },
          start = {
            character = 7,
            line = 3
          }
        }
      }, {
        children = { {
            detail = "instantiation",
            kind = 2,
            name = "ENTITY_1_1: ENTITY_1",
            range = {
              ["end"] = {
                character = 3,
                line = 41
              },
              start = {
                character = 2,
                line = 35
              }
            },
            selectionRange = {
              ["end"] = {
                character = 3,
                line = 41
              },
              start = {
                character = 2,
                line = 35
              }
            }
          }, {
            detail = "instantiation",
            kind = 2,
            name = "ENTITY_1_2: ENTITY_1",
            range = {
              ["end"] = {
                character = 3,
                line = 49
              },
              start = {
                character = 2,
                line = 43
              }
            },
            selectionRange = {
              ["end"] = {
                character = 3,
                line = 49
              },
              start = {
                character = 2,
                line = 43
              }
            }
          }, {
            detail = "instantiation",
            kind = 2,
            name = "ENTITY_2_1: ENTITY_2",
            range = {
              ["end"] = {
                character = 3,
                line = 57
              },
              start = {
                character = 2,
                line = 51
              }
            },
            selectionRange = {
              ["end"] = {
                character = 3,
                line = 57
              },
              start = {
                character = 2,
                line = 51
              }
            }
          }, {
            detail = "instantiation",
            kind = 2,
            name = "ENTITY_2_2: ENTITY_2",
            range = {
              ["end"] = {
                character = 3,
                line = 65
              },
              start = {
                character = 2,
                line = 59
              }
            },
            selectionRange = {
              ["end"] = {
                character = 3,
                line = 65
              },
              start = {
                character = 2,
                line = 59
              }
            }
          }, {
            children = {},
            detail = "process",
            kind = 12,
            name = "PROC_p",
            range = {
              ["end"] = {
                character = 13,
                line = 81
              },
              start = {
                character = 2,
                line = 67
              }
            },
            selectionRange = {
              ["end"] = {
                character = 8,
                line = 67
              },
              start = {
                character = 2,
                line = 67
              }
            }
          }, {
            children = {},
            detail = "process",
            kind = 12,
            name = "no label",
            range = {
              ["end"] = {
                character = 13,
                line = 98
              },
              start = {
                character = 2,
                line = 84
              }
            },
            selectionRange = {
              ["end"] = {
                character = 13,
                line = 98
              },
              start = {
                character = 2,
                line = 84
              }
            }
          } },
        detail = "architecture",
        kind = 5,
        name = "ENTITY_TOP",
        range = {
          ["end"] = {
            character = 17,
            line = 100
          },
          start = {
            character = 0,
            line = 12
          }
        },
        selectionRange = {
          ["end"] = {
            character = 31,
            line = 12
          },
          start = {
            character = 21,
            line = 12
          }
        }
      } }
  }
}
```

Those are the tree different Neovim GUIs when using the three LSP servers:

`rust_hdl`:

image

`VHDL-Tool`:

image

`vhdl-linter`:

image

As you can see the [breadcrumb](https://code.visualstudio.com/docs/editor/editingevolved), which is displayed right at the top of buffer, is not correctly generated with `rust_hdl`.

Also, it seems that symbol scope information is missing in `rust_hdl`. You can see how the symbols panel report a yellow highlight for a node (symbol) and all of its parents (scopes).
This means that when the cursor is placed somewhere inside the buffer, then the curren scope is detected up to the highest-level parent node.

It is important to observe that when using other LSP servers for different languages, for example LUA or CPP, the breadcrumb and symbol scope information are always correctly detected, but that's not the case with `rust_hdl`.

In conclusion, it seems that in its current implementation, `rust_hdl` produces an output in a different format than the other LSPs, not allowing to make use of the breadcrumb and symbol scope features.

### PR DESCRIPTION
**Pull request by Schottkyc137**
Closes #158 

Now adds a source span to all named entity declarations.
Also adds unnamed statements and declarations (i.e., process statements) to the entity hierarchy (visible, for example, in the outline in VHDL).

Any feedback concerning the format is greatly appreciated.
For example, I'm unsure whether `if` statements and similar should be part of the outline or not.

### CODE REVIEW

**Comment by pidgeon777** [comment]
> For example, I'm unsure whether if statements and similar should be part of the outline or not.

Do you think it would be possible to let the user decide what should be included in the outline and what not? Maybe by using the LSP server settings during init, or the project config file?

Or, all of the basic VHDL `if`, `for` etc. tokens could be output, and then filtered out eventually client side.

I think it would be great to also report the `when <value>` of the `case <name> is` in the VHDL outline (useful to quickly jump to FSM states, for example).

**Comment by Schottkyc137** [comment]
> > For example, I'm unsure whether if statements and similar should be part of the outline or not.
> 
> Do you think it would be possible to let the user decide what should be included in the outline and what not? Maybe by using the LSP server settings during init, or the project config file?
> 
> Or, all of the basic VHDL `if`, `for` etc. tokens could be output, and then filtered out eventually client side.
> 

I'm not against a generic configuration. If you could provide a 'user perspective', i.e., how a user would like to configure the various options, I might include it if it isn't too much effort. That being said, such a feature isn't high on my priority list. Therefore I won't give any guarantees that this will actually be implemented. Probably this would belong to the initial server configuration as this is quite specific to the language server.

> I think it would be great to also report the `when <value>` of the `case <name> is` in the VHDL outline (useful to quickly jump to FSM states, for example).

I see the benefit in such a feature, but I won't implement it now. At the moment, the outline only contains named object, i.e., entities, architectures, e.t.c. This also includes objects that _could_ have a name in form of a label, such as if or case statements. `when` clauses do not fall into this category and would thus potentially require a more substantial refactor of the current design.

**Comment by Schottkyc137** [comment]
@kraigher I'm sort of at a crossroad here. The source location of an `AnyEnt` is currently implemented using a `TokenSpan` object while the declaration position is implemented using a `SrcPos` object. I would like to make this more consistent so that either token information or source code location is given (i.e., either exchanging the declaration position with a single `TokenId` or exchanging the whole span with a single `SrcPos` location).
Having the `SrcPos` on an entity has the advantage of being easier, since no context is required to get the actual location, but may be less flexible since finding the original token span is more cumbersome. Having token information could be advantageous however, for example, for extracting documentation during hover. Do you have a preference how this should be implemented?

**Comment by kraigher** [comment]
The reason to move in direction of TokenSpans was to enable better access to comments and lossless formatting. So it makes sense to keep going in this direction.

Using the context to get tokens is not so bad. Having a context on the side also means less information has to be stored in the data structure themselves which is more space efficient.

File: vhdl_lang/src/analysis/sequential.rs [hunk 1]
```diff
@@ -293,7 +298,11 @@ impl<'a, 't> AnalyzeContext<'a, 't> {
                 )?;
             }
             SequentialStatement::VariableAssignment(ref mut assign) => {
-                let VariableAssignment { target, rhs } = assign;
+                let VariableAssignment {
+                    target,
+                    rhs,
+                    // span: _,
```
Comment (Schottkyc137) [inline vhdl_lang/src/analysis/sequential.rs:304]: remove code comment

File: vhdl_lang/src/ast/search.rs [hunk 2]
```diff
@@ -373,7 +373,11 @@ impl Search for LabeledSequentialStatement {
                 return_if_found!(search_assignment(target, rhs, searcher, ctx));
             }
             SequentialStatement::VariableAssignment(ref assign) => {
-                let VariableAssignment { target, rhs } = assign;
+                let VariableAssignment {
+                    target,
+                    rhs,
+                    // span: _,
```
Comment (Schottkyc137) [inline vhdl_lang/src/ast/search.rs:379]: remove code comment

File: vhdl_lang/src/named_entity.rs [hunk 3]
```diff
@@ -192,7 +194,10 @@ pub struct AnyEnt<'a> {
     pub designator: Designator,
     pub kind: AnyEntKind<'a>,
     pub decl_pos: Option<SrcPos>,
-    pub src_span: Option<TokenSpan>,
+    // TODO: There needs to be source information.
```
Comment (Schottkyc137) [inline vhdl_lang/src/named_entity.rs:197]: remove todo comments

File: vhdl_lang/src/syntax/sequential_statement.rs [hunk 4]
```diff
@@ -417,42 +418,50 @@ fn parse_assignment_or_procedure_call(
     ctx: &mut ParsingContext<'_>,
     target: WithTokenSpan<Target>,
 ) -> ParseResult<SequentialStatement> {
+    let target_span = target.span();
     Ok(expect_token!(
         ctx.stream,
         token,
         ColonEq => {
+            let rhs = parse_variable_assignment_right_hand(ctx)?;
```
Comment (Schottkyc137) [inline vhdl_lang/src/syntax/sequential_statement.rs:426]: unnecessary change (also down)
